# Домашнее задание 6. Pandas

**Примечание:**

В заданиях вам даны подсказки по выполнению, a также вывод работы кода авторского решения - вы можете смотреть на этот вывод и понимать, в правильном ли направлении вы движетесь при выполнении задания.

Скачиваем данные:

In [ ]:
!curl -L -o 'articles.csv' 'https://drive.google.com/uc?id=1nht5Pd7kuqJSCz7B9jwU4qScnCO1ZPop&export=download&confirm=t'

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  2424  100  2424    0     0   3485      0 --:--:-- --:--:-- --:--:-- 15843


In [ ]:
import pandas as pd

# Задача 0

Загрузите таблицу с данными из `articles.csv`:


In [ ]:
articles = pd.read_csv('articles.csv')

Посмотрите на первые 5 строк датафрейма (метод `head`):

In [ ]:
articles.head(5)

```
	id	title	publication	author	date	year	month	content
0	17283	House Republicans Fret About Winning Their Hea...	New York Times	Carl Hulse	2016-12-31	2016	12	WASHINGTON — Congressional Republicans have...
1	17284	Rift Between Officers and Residents as Killing...	New York Times	Benjamin Mueller and Al Baker	2017-06-19	2017	6	After the bullet shells get counted, the blood...
2	17285	Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...	New York Times	Margalit Fox	2017-01-06	2017	1	When Walt Disney’s “Bambi” opened in 1942, cri...
3	17286	Among Deaths in 2016, a Heavy Toll in Pop Musi...	New York Times	William McDonald	2017-04-10	2017	4	Death may be the great equalizer, but it isn’t...
4	17287	Kim Jong-un Says North Korea Is Preparing to T...	New York Times	Choe Sang-Hun	2017-01-02	2017	1	SEOUL, South Korea — North Korea’s leader, ...
```

Определите количество строк и столбцов в датафрейме (атрибут `shape`)

In [ ]:
articles.shape

`(50000, 8)`

Посмотрите, в каких столбцах есть пропуски (используйте методы ```isnull``` и ```sum```)

In [ ]:
articles.isnull().sum()

```
id                0
title             0
publication       0
author         6306
date              0
year              0
month             0
content           0
dtype: int64
```

Видим, что в колонке ```author``` есть пропуски.

Посчитайте, какая доля пропусков содержится в этой колонке (используйте методы ```isnull``` и ```mean```)

In [ ]:
articles['author'].isnull().mean()

`0.12612`

Видим, что в колонке ```author``` 12% пропусков. Давайте удалим те строки, в которых пропущено значение в колонке ```author```

Для этого используйте функцию ```dropna```, которой в качестве аргумента можно передать параметр ```axis```, чтобы удалить строки или столбцы с пропусками.

Не забудьте, что чтобы сохранить результат удаления строк, нужно либо перезаписать датафрейм (```df = df.same_method()```), либо выставить в методе ```dropna``` параметр ```inplace=True```, который говорит о том, что операция будет производиться с самим датафреймом, а не с его копией


In [ ]:
articles.dropna?

ValueError: The truth value of a DataFrame is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [ ]:
articles.dropna(inplace=True)

Проверим размер датафрейма после удаления пропусков и убедимся, что в датафрейме теперь нет пропусков в данных

In [ ]:
articles.shape

(43694, 8)

`(43694, 8)`

In [ ]:
articles.isnull().sum()

id             0
title          0
publication    0
author         0
date           0
year           0
month          0
content        0
dtype: int64

```
id             0
title          0
publication    0
author         0
date           0
year           0
month          0
content        0
dtype: int64
```

Проделайте следующие базовые операции с датафреймами:

# Задача 1


Определите количество различных издательств в таблице и их имена (используйте для этого методы датафрейма ```unique``` или ```nunique```):

In [ ]:
articles['publication'].nunique()

5

`5`

In [ ]:
articles['publication'].unique()

array(['New York Times', 'Breitbart', 'CNN', 'Business Insider',
       'Atlantic'], dtype=object)

```
array(['New York Times', 'Breitbart', 'CNN', 'Business Insider',
       'Atlantic'], dtype=object)
```

# Задача 2

Найдите количество опубликованных статей в отрезке с ```2016-06-03``` по ```2016-12-07``` (включая эти дни)

Для этого используйте фильтрацию по условию. С датами можно обращаться просто как со строками.

При желании это все можно написать одной строкой.

**Подсказка:** найдите строки c датой публикации ```>=``` стартовой даты и получите булеву маску. Аналогично получите булеву маску для второй даты. Затем с помощью операции ```&``` над двумя булевыми масками получите новый булев массив, в котором ```True``` будет стоять только у тех строк, у которых дата публикации попадает в желаемый отрезок. А затем используйте эту булеву маску как индексы датафрейма.

In [ ]:
articles[(articles['date']>= '2016-06-03') & (articles['date']<= '2016-12-07')].shape[0]

15079

`15079`

# Задача 3

Какое издание выпустило больше всего статей в 2016 году?

In [ ]:
a = articles[articles['year'] ==2016]

#b=a.groupby('publication').count()[['id']]
#b[b['id']== b['id'].max()]
a['publication'].value_counts().head(1)


Breitbart    16100
Name: publication, dtype: int64

`'Breitbart'`

# Задача 4

Посчитайте количество статей автора *Tom Ciccotta* в каждом году.

Для этого сначала отберите строки, в которых автором является *Tom Ciccotta*, а затем создайте счетчик из столбца ```year``` (для этого просто позовите метод ```value_counts()``` от столбца ```year```)

In [ ]:
articles[articles['author']=='Tom Ciccotta']['year'].value_counts()

2017    132
2016    124
Name: year, dtype: int64

```
2017    132
2016    124
Name: year, dtype: int64
```

## Задача 5

Выпишите названия 3х первых статей автора *John Hayward* в 2016 году.

**Подсказка 1:** после отбора статей по автору и году, не забудте отсортировать датафрейм по столбцу даты публикации.

**Подсказка 2:** срезы по строкам датафреймов работают так же, как срезы по спискам.

In [ ]:
a = articles[(articles['author']=='John Hayward') & (articles['year']==2016)].sort_values(by='date')
a.iloc[0:3]['title']

16861    Indonesian Couple Beaten with Canes for Violat...
22041    Islamic State Claims Credit For Gun Attack On ...
27561    Protests Across India Against Saudi Execution ...
Name: title, dtype: object

```
array(['Indonesian Couple Beaten with Canes for Violating Sharia Law',
       'Islamic State Claims Credit For Gun Attack On Russian Tourists',
       'Protests Across India Against Saudi Execution of Shiite Cleric'],
      dtype=object)
```

# Задача 6

Какой автор на сотом месте по количеству опубликованных статей? Сколько статей он опубликовал?

**Подсказка:** для подсчета встречаемости значений используйте метод ```value_counts```. Для доступа к индексам полученного после ```value_counts``` ```pd.Series``` используйте атрибут ```index```

In [ ]:
a=articles['author'].value_counts()
a.index.to_list()[100-1]

'Milo'

`'Milo'`

In [ ]:
# число статей
a = articles[articles['author'] == 'Milo']
a.shape[0]

86

`86`

# Работа со строками в датафрейме

Еще одной важной частью работы с данными является **нормализация данных**.

Это особенно актуально при работе с данными в текством формате и при работе с данными из нескольких разнородных источников. Например, один и тот же номер телефона пользователя может записан в базах данных в разных форматах:

- `+7(918)123-45-67`
- `+7(918)1234567`
- `+79181234567`
- `+8(918)123-45-67`
- `8(918)1234567`
- `89181234567`

Поэтому важно нормализовывать данные, чтобы иметь возможность объединить все данные, которые относятся к одному реальному объекту.

Именно нормализации данных при работе со строками будем учиться в этой части домашней работы.

Для датафреймов существуют методы работы со строковыми данными. Чтобы применить их, необходимо воспользоваться атрибутом ```str```, после чего вызвать нужные методы работы со строками. Например, вызов:

```df['content'].str.len() ```

подсчитает для каждой строчки в датафрейме количество символов в колонке content. Примеры кода по работе с текстовыми данными в pandas можно найти в гайде [working with text data](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html) документации, там же вы найдете список методов пандаса по работе с текстовыми данными.


# Задача 7

Найдите в датафрейме всех авторов, имя которых содержит ```Faith```. Выведите Series, состоящий из всех таких уникальных имен.

In [ ]:
a = articles[articles['author'].str.contains('Faith',na= False)]
a['author'].unique()

array([' Faith Haleh Robinson', 'Faith Karimi', 'Faith Haleh Robinson',
       'Faith Karimi,', 'Faith Karimi ', 'Faith Karimi, ',
       ' Faith Karimi', ' Faith Karimi '], dtype=object)

```
0     Faith Haleh Robinson
1             Faith Karimi
2     Faith Haleh Robinson
3            Faith Karimi,
4            Faith Karimi
5           Faith Karimi,
6             Faith Karimi
7            Faith Karimi
dtype: object
```

Как можно заметить, в таблице существует множество различных написаний имени Faith Karimi. В основном эти написания различаются пунктуацией - лишние пробелы и запятые. Для правильного подсчета статистик для текстовых данных зачастую возникает необходимость в их предобработке.

**Методы строк** можно вспомнить по этой ссылке: https://pythonworld.ru/tipy-dannyx-v-python/stroki-funkcii-i-metody-strok.html

Проведите следующие преобразования для колонок ```author``` и ```content```:

# Задача 8

Приведение текста к нижнему регистру:

In [ ]:
new_author = articles['author'].str.lower()
articles['author'] = new_author
new_content = articles['content'].str.lower()
articles['content'] = new_content
articles.head()

id                                              title     publication  \
0  17283  House Republicans Fret About Winning Their Hea...  New York Times   
1  17284  Rift Between Officers and Residents as Killing...  New York Times   
2  17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...  New York Times   
3  17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...  New York Times   
4  17287  Kim Jong-un Says North Korea Is Preparing to T...  New York Times   

                          author        date  year  month  \
0                     carl hulse  2016-12-31  2016     12   
1  benjamin mueller and al baker  2017-06-19  2017      6   
2                   margalit fox  2017-01-06  2017      1   
3               william mcdonald  2017-04-10  2017      4   
4                  choe sang-hun  2017-01-02  2017      1   

                                             content  
0  washington  —   congressional republicans have...  
1  after the bullet shells get counted, the blood...  
2  when walt disney’s “bambi” opened in 1942, cri...  
3  death may be the great equalizer, but it isn’t...  
4  seoul, south korea  —   north korea’s leader, ...

```
     id	title	publication	author	date	year	month	content
0	17283	House Republicans Fret About Winning Their Hea...	New York Times	carl hulse	2016-12-31	2016	12	washington — congressional republicans have...
1	17284	Rift Between Officers and Residents as Killing...	New York Times	benjamin mueller and al baker	2017-06-19	2017	6	after the bullet shells get counted, the blood...
2	17285	Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...	New York Times	margalit fox	2017-01-06	2017	1	when walt disney’s “bambi” opened in 1942, cri...
3	17286	Among Deaths in 2016, a Heavy Toll in Pop Musi...	New York Times	william mcdonald	2017-04-10	2017	4	death may be the great equalizer, but it isn’t...
4	17287	Kim Jong-un Says North Korea Is Preparing to T...	New York Times	choe sang-hun	2017-01-02	2017	1	seoul, south korea — north korea’s leader, ...
```

# Задача 9

Удаление всей пунктуации из текста.

**Подсказка:** для вас уже создано начальное множество символов (достаточно ли его?). Символы надо исключить с помощью метода ```replace```.

Данного множества недостаточно, для примера, не хватает следующих символов: ’, ” и тд.

In [ ]:
import string
punctuation = set(string.punctuation)

In [ ]:
# может работать несколько секунд
articles['content'] = articles['content'].str.replace('['+string.punctuation+']', '', regex=True)
articles['content'] = articles['content'].str.replace('-\,"', '')
articles['content'] = articles['content'].replace('(,$)', '', regex = True)
articles['author'] = articles['author'].str.replace('['+string.punctuation+']', '', regex=True)
articles

C:\Users\adamc\AppData\Local\Temp\ipykernel_6068\2125963329.py:3: FutureWarning: The default value of regex will change from True to False in a future version.
  articles['content'] = articles['content'].str.replace('-\,"', '')


id                                              title  \
0      17283  House Republicans Fret About Winning Their Hea...   
1      17284  Rift Between Officers and Residents as Killing...   
2      17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...   
3      17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...   
4      17287  Kim Jong-un Says North Korea Is Preparing to T...   
...      ...                                                ...   
49995  73465   Rex Tillerson Says Climate Change Is Real, but …   
49996  73466  The Biggest Intelligence Questions Raised by t...   
49997  73467  Trump Announces Plan That Does Little to Resol...   
49998  73468    Dozens of For-Profit Colleges Could Soon Close    
49999  73469                       The Milky Way’s Stolen Stars   

          publication                         author        date  year  month  \
0      New York Times                     carl hulse  2016-12-31  2016     12   
1      New York Times  benjamin mueller and al baker  2017-06-19  2017      6   
2      New York Times                   margalit fox  2017-01-06  2017      1   
3      New York Times               william mcdonald  2017-04-10  2017      4   
4      New York Times                   choe sanghun  2017-01-02  2017      1   
...               ...                            ...         ...   ...    ...   
49995        Atlantic                 robinson meyer  2017-01-11  2017      1   
49996        Atlantic                     amy zegart  2017-01-11  2017      1   
49997        Atlantic                  jeremy venook  2017-01-11  2017      1   
49998        Atlantic                    emily deruy  2017-01-11  2017      1   
49999        Atlantic                   marina koren  2017-01-11  2017      1   

                                                 content  
0      washington — congressional republicans have a ...  
1      after the bullet shells get counted the blood ...  
2      when walt disney’s “bambi” opened in 1942 crit...  
3      death may be the great equalizer but it isn’t ...  
4      seoul south korea — north korea’s leader kim s...  
...                                                  ...  
49995  as chairman and ceo of exxonmobil rex tillerso...  
49996  i’ve spent nearly 20 years looking at intellig...  
49997  donald trump will not be taking necessary step...  
49998  dozens of colleges could be forced to close in...  
49999  the force of gravity can be described using a ...  

[50000 rows x 8 columns]

In [ ]:
# пример того, что должно получиться после этого шага
articles.content.loc[45926]

'”back in the late ’90s, getting a startup big enough to get onto the stock market was the thing to do. ” ’now, a market crash, a great recession, and a possible   burst later  —   investors are  and the public stock market is increasingly volatile.’ ’still, going public is often viewed as the end goal for ambitious young tech companies. ’ ”here’s how and why startups wind up on the stock market  —   and why .” ”back in the late ’90s, getting a startup big. ..”'

```
'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do  now a market crash a great recession and a possible burst later  investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies  heres how and why startups wind up on the stock market  and why  back in the late 90s getting a startup big '
```

# Задача 10

Удаление пробелов в начале и конце строки (воспользуйтесь готовым методом):

In [ ]:
new_author = articles['author'].str.strip()
articles['author'] = new_author
new_content = articles['content'].str.strip()
articles['content'] = new_content
articles.head()

id                                              title     publication  \
0  17283  House Republicans Fret About Winning Their Hea...  New York Times   
1  17284  Rift Between Officers and Residents as Killing...  New York Times   
2  17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...  New York Times   
3  17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...  New York Times   
4  17287  Kim Jong-un Says North Korea Is Preparing to T...  New York Times   

                          author        date  year  month  \
0                     carl hulse  2016-12-31  2016     12   
1  benjamin mueller and al baker  2017-06-19  2017      6   
2                   margalit fox  2017-01-06  2017      1   
3               william mcdonald  2017-04-10  2017      4   
4                   choe sanghun  2017-01-02  2017      1   

                                             content  
0  washington — congressional republicans have a ...  
1  after the bullet shells get counted the blood ...  
2  when walt disney’s “bambi” opened in 1942 crit...  
3  death may be the great equalizer but it isn’t ...  
4  seoul south korea — north korea’s leader kim s...

In [ ]:
# check
articles.content.loc[45926]

'”back in the late ’90s getting a startup big enough to get onto the stock market was the thing to do ” ’now a market crash a great recession and a possible   burst later  —   investors are  and the public stock market is increasingly volatile’ ’still going public is often viewed as the end goal for ambitious young tech companies ’ ”here’s how and why startups wind up on the stock market  —   and why ” ”back in the late ’90s getting a startup big ”'

```
'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do  now a market crash a great recession and a possible burst later  investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies  heres how and why startups wind up on the stock market  and why  back in the late 90s getting a startup big'
```

# Задача 11

Замена подряд идущих пробелов одним пробелом:

In [ ]:
articles['content'] = articles['content'].str.replace('  ', ' ')
articles['author'] = articles['author'].str.replace('  ', ' ')
articles.head()

id                                              title     publication  \
0  17283  House Republicans Fret About Winning Their Hea...  New York Times   
1  17284  Rift Between Officers and Residents as Killing...  New York Times   
2  17285  Tyrus Wong, ‘Bambi’ Artist Thwarted by Racial ...  New York Times   
3  17286  Among Deaths in 2016, a Heavy Toll in Pop Musi...  New York Times   
4  17287  Kim Jong-un Says North Korea Is Preparing to T...  New York Times   

                          author        date  year  month  \
0                     carl hulse  2016-12-31  2016     12   
1  benjamin mueller and al baker  2017-06-19  2017      6   
2                   margalit fox  2017-01-06  2017      1   
3               william mcdonald  2017-04-10  2017      4   
4                   choe sanghun  2017-01-02  2017      1   

                                             content  
0  washington — congressional republicans have a ...  
1  after the bullet shells get counted the blood ...  
2  when walt disney’s “bambi” opened in 1942 crit...  
3  death may be the great equalizer but it isn’t ...  
4  seoul south korea — north korea’s leader kim s...

In [ ]:
# check
articles.content.loc[45926]

```
'back in the late 90s getting a startup big enough to get onto the stock market was the thing to do now a market crash a great recession and a possible burst later investors are and the public stock market is increasingly volatile still going public is often viewed as the end goal for ambitious young tech companies heres how and why startups wind up on the stock market and why back in the late 90s getting a startup big'
```

Если последовательно применить код из задач 8-11 к строке ```"   It's 6 a.m. and I'm still doing this homework :((    "```, то она преобразуется в строку  ```its 6 am and im still doing this homework```


Снова найдите в датафрейме всех авторов, имя которых содержит faith. Проверьте, что теперь различные способы написания «схлопываются» в один.

In [ ]:
a = articles[articles['author'].str.contains('faith',na= False)]
a['author'].unique()

array(['faith haleh robinson', 'faith karimi'], dtype=object)

```
0    faith haleh robinson
1            faith karimi
dtype: object
```

Еще задачи:

- https://www.machinelearningplus.com/python/101-pandas-exercises-python/
- https://www.kaggle.com/code/tangchengshun/pandas-75-exercises-with-solutions
- https://pandaspractice.com/
- https://www.w3resource.com/python-exercises/pandas/index.php
- https://grossmend.com/blog/post/pandas_120_part_2/